### ISSUE TITLE
Update Lifecycle Hooks for Future Async Rendering

### ISSUE THREAD

**Issue by ryanbrainard** [heroku/react-refetch#214]
We will need to this:

https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html#fetching-external-data

ideally before a "future 16.x release" and definitely before 17.0

**Comment by villesau** [comment]
@ryanbrainard This prevents us from using e.g `getDerivedStateFromProps` with react-refetch. Using new lifecycle methods means that old lifecycle methods are not called:

```
vendor.js:504207 Warning: Unsafe legacy lifecycles will not be called for components using new component APIs.

Refetch.connect(MyComponent) uses getDerivedStateFromProps() but also contains the following legacy lifecycles:
  componentWillMount
  componentWillReceiveProps
  componentWillUpdate
```

**Comment by ryanbrainard** [comment]
Merged in https://github.com/heroku/react-refetch/issues/231. It looks like https://github.com/heroku/react-refetch/pull/217 has been abandoned, but I am going to see if I can dust things off and move forward with this.

**Comment by ryanbrainard** [comment]
Released in [3.0.0-1](https://github.com/heroku/react-refetch/releases/tag/v3.0.0-1 Beta)
